File: haskell/nucleotide-count/DNA.hs
{-# LANGUAGE TupleSections #-}
module DNA
    ( count
    , nucleotideCounts
    ) where

import Data.Map.Strict ( Map, (!), fromListWith )

type Nucleotide = Char

dna :: [Nucleotide]
dna = "ACGT"

count :: Nucleotide -> [Nucleotide] -> Int
count nuc strand 
  | nuc `elem` dna = nucleotideCounts strand ! nuc
  | nuc == 'U'     = 0
  | otherwise      = error $ "invalid nucleotide " ++ show nuc

nucleotideCounts :: [Nucleotide] -> Map Nucleotide Int
nucleotideCounts = fromListWith (+) . (++ zeros) . map (,1)
 where zeros = map (,0) dna

File: go/nucleotide-count/dna.go
package dna

import "fmt"

type DNA string

type Histogram map[byte]int

func (dna DNA) Counts() Histogram {
	h := Histogram{'A': 0, 'C': 0, 'T': 0, 'G': 0}
	for i := 0; i < len(dna); i++ {
		h[dna[i]]++
	}
	return h
}

func (dna DNA) Count(n byte) (int, error) {
	if count, ok := dna.Counts()[n]; ok {
		return count, nil
	} else if n == 'U' {
		return 0, nil
	} else {
		return 0, fmt.Errorf("%[1]q.Count(%[2]q): invalid nucleotide %[2]q", dna, n)
	}
}
